### PR TITLE
#1512 notification failures table

### PR DESCRIPTION
**Pull request by nikolai-efimov**
# Description

Because we are validating certain items after returning a 202 Accepted from the v3 POST routes, we need to be able to handle validation failures for certain instances with a new table in PostgreSQL

issue [#1512](https://app.zenhub.com/workspaces/va-notify-620d21369d810a00146ed9c8/issues/gh/department-of-veterans-affairs/notification-api/1512)

## Type of change

Please check the relevant option(s).

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [ ] Documentation changes only

## How Has This Been Tested?

Migrations (up and down) were run locally to verify the change. 

Screen Shot 2023-11-01 at 5 50 38 PM

Screen Shot 2023-11-01 at 5 54 40 PM

## Checklist

- [x] I have assigned myself to this PR
- [x] PR has an appropriate title: #9999 - What the thing does
- [ ] PR has a detailed description, including links to specific documentation
- [x] I have added the [appropriate labels](https://github.com/department-of-veterans-affairs/notification-api/blob/master/.github/release.yaml) to the PR.
- [x] I did not remove any parts of the template, such as checkboxes even if they are not used
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to any documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have ensured the latest master is merged into my branch and all checks are green prior to review
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [x] The ticket is now moved into the DEV test column

### CODE REVIEW

File: app/models.py [hunk 1]
```diff
@@ -2218,3 +2218,26 @@ def serialize(self):
...
+        created_at (DateTime): Record date and time of creation. This field will be used for pg_cron
+        job delete records older then 30 days
+        body (JSONB): Column used to store the details of the notification failure in a JSON object.
+
+    Methods:
+        serialize: Converts the 'body' attribute to a dictionary for easier consumption in Python.
+    """
+    __tablename__ = 'notification_failures'
+
+    notification_id = db.Column(db.Integer, primary_key=True)
```
Comment (cris-oddball) [inline app/models.py:2238]: Not understanding how this is an integer when a notification_id is always a UUID? Is that how the notifications table is configured?

**Review by cris-oddball — CHANGES_REQUESTED** [review]
This should be deployed to Dev and tested there, please.

File: app/models.py [hunk 2]
```diff
@@ -2218,3 +2218,26 @@ def serialize(self):
             'created_at': self.created_at.strftime(DATETIME_FORMAT),
             'updated_at': self.updated_at.strftime(DATETIME_FORMAT) if self.updated_at else None
         }
+
+
+class NotificationFailures(db.Model):
+    """
+    A SQLAlchemy model representing the 'notification_failures' table.
```
Comment (k-macmillan) [inline app/models.py:2225]:
```suggestion
    A SQLAlchemy model representing the 'notification_failures' table. 
    This table captures infrequent, invalid v3 requests that passed schema checks but failed during processing and could not be added to the notifications table.
```

Something like that. We want to capture the purpose of the table and an extra sentence can prevent a future engineer from having to dig through docs to understand why this exists.

File: app/models.py (hunk 1 above)
Comment (k-macmillan) [inline app/models.py:2238]: I don't believe that should be `db.Integer`. This is a UUID field.

File: migrations/versions/0365_add_notification_failures.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+revision = "0365_add_notification_failures"
+down_revision = "0364_add_templatep2pchecklist"
+
+
+def upgrade():
+    op.create_table(
+        "notification_failures",
+        sa.Column("notification_id", postgresql.UUID(as_uuid=True)),
+        sa.Column("body", postgresql.JSONB(astext_type=sa.Text()), nullable=True),
```
Comment (k-macmillan) [inline migrations/versions/0365_add_notification_failures.py:20]: > astext_type=sa.Text())

What was the purpose of this? Does this turn integers into text for comparison purposes? I would have expected it kept the appropriate data type.

**Review by k-macmillan — CHANGES_REQUESTED** [review]

**Review by cris-oddball — APPROVED** [review]

File: app/models.py (hunk 1 above)
Comment (nikolai-efimov) [reply app/models.py]: done

File: app/models.py (hunk 2 above)
Comment (nikolai-efimov) [reply app/models.py]: done

File: app/models.py (hunk 1 above)
Comment (nikolai-efimov) [reply app/models.py]: done

File: migrations/versions/0365_add_notification_failures.py (hunk 3 above)
Comment (nikolai-efimov) [reply migrations/versions/0365_add_notification_failures.py]: done

**Pull request merged**